core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala

```diff
@@ -232,22 +232,23 @@ private [util] class SparkShutdownHookManager {
     }
   }
 
-  def runAll(): Unit = synchronized {
+  def runAll(): Unit = {
     shuttingDown = true
-    while (!hooks.isEmpty()) {
-      Try(Utils.logUncaughtExceptions(hooks.poll().run()))
+    var nextHook: SparkShutdownHook = null
+    while ({nextHook = hooks synchronized { hooks.poll() }; nextHook != null}) {
+      Try(Utils.logUncaughtExceptions(nextHook.run()))
     }
   }
 
-  def add(priority: Int, hook: () => Unit): AnyRef = synchronized {
+  def add(priority: Int, hook: () => Unit): AnyRef = {
     checkState()
     val hookRef = new SparkShutdownHook(priority, hook)
-    hooks.add(hookRef)
+    hooks synchronized { hooks.add(hookRef) }
```

Comment:
Yeah, you got me. That really has to be tightened up. Will address these shortly.